gdnative-core/src/core_types/quat.rs

```diff
@@ -9,22 +10,274 @@ pub struct Quat {
+        Basis::from_quat(self).to_euler()
+    }
+
+    /// Returns the inverse of the quaternion.
+    #[inline]
+    pub fn inverse(self) -> Self {
+        Self::gd(self.glam().inverse())
+    }
+
+    /// Returns `true` if this quaternion and `quat` are approximately equal, by running
+    /// `is_equal_approx` on each component
     #[inline]
-    pub fn is_equal_approx(self, to: &Self) -> bool {
+    pub fn is_equal_approx(self, to: Self) -> bool {
         self.x.is_equal_approx(to.x)
             && self.y.is_equal_approx(to.y)
             && self.z.is_equal_approx(to.z)
             && self.w.is_equal_approx(to.w)
     }
 
+    /// Returns whether the quaternion is normalized or not.
+    #[inline]
+    pub fn is_normalized(self) -> bool {
+        self.glam().is_normalized()
+    }
+
+    /// Returns the length of the quaternion.
+    #[inline]
+    pub fn length(self) -> f32 {
+        self.glam().length()
+    }
+
+    /// Returns the length of the quaternion, squared.
+    #[inline]
+    pub fn length_squared(self) -> f32 {
+        self.glam().length_squared()
+    }
+
+    /// Returns a copy of the quaternion, normalized to unit length.
     #[inline]
-    fn glam(self) -> glam::Quat {
+    pub fn normalized(self) -> Self {
+        Self::gd(self.glam().normalize())
+    }
+
+    /// Returns the result of the spherical linear interpolation between this quaternion and to by
+    /// amount weight.
+    ///
+    /// **Note:** Both quaternions must be normalized.
+    #[inline]
+    pub fn slerp(self, b: Self, t: f32) -> Self {
+        debug_assert!(self.is_normalized(), "Quaternion `self` is not normalized");
+        debug_assert!(b.is_normalized(), "Quaternion `b` is not normalized");
+
+        // Copied from Godot's Quat::slerp as glam::lerp version diverges too much
+
+        // calc cosine
+        let cosom = self.dot(b);
+
+        // adjust signs (if necessary)
+        let (cosom, b) = if cosom < 0.0 {
+            (-cosom, -b)
+        } else {
+            (cosom, b)
+        };
+
+        // calculate coefficients
+        let scale = if (1.0 - cosom) > CMP_EPSILON as f32 {
+            // standard case (slerp)
+            let omega = cosom.acos();
+            let sinom = omega.sin();
+            (((1.0 - t) * omega).sin() / sinom, (t * omega).sin() / sinom)
+        } else {
+            // "from" and "to" quaternions are very close
+            //  ... so we can do a linear interpolation
+            (1.0 - t, t)
+        };
+
+        // calculate final values
+        Self::new(
+            scale.0 * self.x + scale.1 * b.x,
+            scale.0 * self.y + scale.1 * b.y,
+            scale.0 * self.z + scale.1 * b.z,
+            scale.0 * self.w + scale.1 * b.w,
+        )
+    }
+
+    /// Returns the result of the spherical linear interpolation between this quaternion and `t` by
+    /// amount `t`, but without checking if the rotation path is not bigger than 90 degrees.
+    #[inline]
```

Comment:
Should this really be inline?

Reply:
Clippy throws an error if it isn't marked as inline. IMO it doesn't need to be inline perse since the generated assembly is quite large, so I could perhaps add `#[allow(clippy::missing_inline_in_public_items)]`

Reply:
Ah, I see. Have we configured it to make _every_ public function `#[inline]`? I assume it would do so based on some heuristic?

I'm fine either way (leaving inline or the allow). Might need to look at this in general, but that's out of scope for the PR.

Reply:
From what I can tell the lint just checks if the function is public: https://rust-lang.github.io/rust-clippy/master/#missing_inline_in_public_items.

Reply:
From that link:

> **What it does**
> it lints if an exported function, method, trait method with default impl, or trait method impl is not #[inline].
>
> **Why is this bad** 
> In general, it is not. Functions can be inlined across crates when that’s profitable as long as any form of LTO is used. When LTO is disabled, functions that are not #[inline] cannot be inlined across crates. Certain types of crates might intend for most of the methods in their public API to be able to be inlined across crates even when LTO is disabled. For these types of crates, enabling this lint might make sense. It allows the crate to require all exported methods to be #[inline] by default, and then opt out for specific methods where this might not make sense.

So looks like it's not a general best practice, but a policy that library authors can opt-in.
Since we already have it enabled, we can add the `#[inline]` for now and maybe rethink this in the future.